neuralpp/inference/graphical_model/representation/factor/symbolic_factor.py

```diff
@@ -0,0 +1,148 @@
+        for a in variable.assignments():
+            symbol = BasicVariable(variable.name, type(a))
+            eq_operator = BasicConstant(operator.eq, Callable[[type(a), type(a)], bool])
+            context = BasicFunctionApplication(eq_operator, [symbol, BasicConstant(a, type(a))])
+            simplified_expression = interpreter.simplify(self.expression, context)
+
+            if result_expression == None:
+                result_expression = simplified_expression
+            else:
+                add_operator = BasicConstant(operator.add, Callable[[Any, Any], Any])
+                result_expression = BasicFunctionApplication(add_operator, [result_expression, simplified_expression])
+
+        result_expression = interpreter.simplify(result_expression)
+        return self.new_instance(result_variables, result_expression)
+
+    def argmax(self):
+        raise NotImplementedError("TODO")
+
+    def normalize(self):
+        interpreter = SymPyInterpreter()
+        sum_expression = self.sum_out_variables(self.variables).expression
+
+        div_operator = BasicConstant(operator.truediv, self._any_function_of_arity([self.expression, sum_expression]))
+        result_expression = BasicFunctionApplication(div_operator, [self.expression, sum_expression])
+        result_expression = interpreter.simplify(result_expression)
+        return self.new_instance(self.variables, result_expression)
+
+    def randomize(self):
+        raise NotImplementedError("TODO")
+
+    def randomized_copy(self):
+         raise NotImplementedError("TODO")
+
+    def __eq__(self, other):
+        """
+        Compares factors by checking if they have the same variables
+        (after appropriate permutation if variables are not in the same order).
+        """
+        if isinstance(other, SymbolicFactor):
+            return (self.variables == other.variables and
+                self.expression == other.expression)
+        else:
+            raise Exception(
+                f"Comparison of SymbolicFactor to factors other than SymbolicFactor is not implemented. "
+                f"Got {type(other)}"
+            )
+
+    def __repr__(self):
+        return "Factor on (" + join(self.variables) + "): " + repr(self.expression)
+
+    def __str__(self):
+        return "Factor on (" + join(self.variables) + "): " + str(self.expression)
+
+    def _variable_type(self, variable: Variable) -> type:
+        if (isinstance(variable, IntegerVariable)):
+            return int
+        else:
+            return float
+
+    def _dict_to_context(self, assignment_dict: dict) -> Expression:
+        result = BasicConstant(True)
+        type_dict = {}
+        for k, v in assignment_dict.items():
+            symbol = BasicVariable(k.name, self._variable_type(v))
+            eq_operator = BasicConstant(operator.eq, Callable[[type(v), type(v)], bool])
+            eq_expression = BasicFunctionApplication(eq_operator, [symbol, BasicConstant(v, type(v))])
```

Comment:
Overloaded operator here too.

Reply:
I'm not quite sure how I can overload `eq` here, since I would need something to contain the expression `symbol = v`. I don't think I can contain that in `BasicVariable` or `BasicConstant`